x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ShrinkAction.java

```diff
@@ -158,11 +182,12 @@ public List<Step> toSteps(Client client, String phase, Step.StepKey nextStepKey)
         StepKey isShrunkIndexKey = new StepKey(phase, NAME, ShrunkenIndexCheckStep.NAME);
         StepKey replaceDataStreamIndexKey = new StepKey(phase, NAME, ReplaceDataStreamBackingIndexStep.NAME);
         StepKey deleteIndexKey = new StepKey(phase, NAME, DeleteStep.NAME);
+        StepKey allowWritesKey = new StepKey(phase, NAME, AllowWritesStep.NAME);
 
         AsyncBranchingStep conditionalSkipShrinkStep = new AsyncBranchingStep(
             preShrinkBranchingKey,
             checkNotWriteIndex,
-            nextStepKey,
+            allowWritesKey,
```

Comment:
@dakrone  pointed out that in the case where shrink does not happen due to the shard count not changing, the settings update in the ShrinkStep would not be run. This is not a problem if the input index is writable, since the ReadOnlyStep is also skipped by the conditional branch. But, if the input index is already read-only (which would be a bit weird), the shrunken index will remain read-only, even if the user set `allowWritesOnTarget=true` . 
To deal with this situation, I've moved the settings update to a separate step which is always run at the end of the ShrinkAction, even when all other steps are skipped.